File: tests/algorithms/x/PHP/maths/quadratic_equations_complex_numbers.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function add($a, $b) {
  return ['re' => $a['re'] + $b['re'], 'im' => $a['im'] + $b['im']];
};
  function sub($a, $b) {
  return ['re' => $a['re'] - $b['re'], 'im' => $a['im'] - $b['im']];
};
  function div_real($a, $r) {
  return ['re' => $a['re'] / $r, 'im' => $a['im'] / $r];
};
  function sqrt_newton($x) {
  if ($x <= 0.0) {
  return 0.0;
}
  $guess = $x / 2.0;
  $i = 0;
  while ($i < 20) {
  $guess = ($guess + $x / $guess) / 2.0;
  $i = $i + 1;
};
  return $guess;
};
  function sqrt_to_complex($d) {
  if ($d >= 0.0) {
  return ['re' => sqrt_newton($d), 'im' => 0.0];
}
  return ['re' => 0.0, 'im' => sqrt_newton(-$d)];
};
  function quadratic_roots($a, $b, $c) {
  if ($a == 0.0) {
  echo rtrim('ValueError: coefficient \'a\' must not be zero'), PHP_EOL;
  return [];
}
  $delta = $b * $b - 4.0 * $a * $c;
  $sqrt_d = sqrt_to_complex($delta);
  $minus_b = ['re' => -$b, 'im' => 0.0];
  $two_a = 2.0 * $a;
  $root1 = div_real(add($minus_b, $sqrt_d), $two_a);
  $root2 = div_real(sub($minus_b, $sqrt_d), $two_a);
  return [$root1, $root2];
};
  function root_str($r) {
  if ($r['im'] == 0.0) {
  return _str($r['re']);
}
  $s = _str($r['re']);
  if ($r['im'] >= 0.0) {
  $s = $s . '+' . _str($r['im']) . 'i';
} else {
  $s = $s . _str($r['im']) . 'i';
}
  return $s;
};
  function main() {
  $roots = quadratic_roots(5.0, 6.0, 1.0);
  if (count($roots) == 2) {
  echo rtrim('The solutions are: ' . root_str($roots[0]) . ' and ' . root_str($roots[1])), PHP_EOL;
}
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/x/Kotlin/maths/quadratic_equations_complex_numbers.kt
fun _numToStr(v: Number): String {
    val d = v.toDouble()
    val i = d.toLong()
    return if (d == i.toDouble()) i.toString() else d.toString()
}

data class Complex(var re: Double = 0.0, var im: Double = 0.0)
fun add(a: Complex, b: Complex): Complex {
    return Complex(re = a.re + b.re, im = a.im + b.im)
}

fun sub(a: Complex, b: Complex): Complex {
    return Complex(re = a.re - b.re, im = a.im - b.im)
}

fun div_real(a: Complex, r: Double): Complex {
    return Complex(re = a.re / r, im = a.im / r)
}

fun sqrt_newton(x: Double): Double {
    if (x <= 0.0) {
        return 0.0
    }
    var guess: Double = x / 2.0
    var i: Int = (0).toInt()
    while (i < 20) {
        guess = (guess + (x / guess)) / 2.0
        i = i + 1
    }
    return guess
}

fun sqrt_to_complex(d: Double): Complex {
    if (d >= 0.0) {
        return Complex(re = sqrt_newton(d), im = 0.0)
    }
    return Complex(re = 0.0, im = sqrt_newton(0.0 - d))
}

fun quadratic_roots(a: Double, b: Double, c: Double): MutableList<Complex> {
    if (a == 0.0) {
        println("ValueError: coefficient 'a' must not be zero")
        return mutableListOf<Complex>()
    }
    var delta: Double = (b * b) - ((4.0 * a) * c)
    var sqrt_d: Complex = sqrt_to_complex(delta)
    var minus_b: Complex = Complex(re = 0.0 - b, im = 0.0)
    var two_a: Double = 2.0 * a
    var root1: Complex = div_real(add(minus_b, sqrt_d), two_a)
    var root2: Complex = div_real(sub(minus_b, sqrt_d), two_a)
    return mutableListOf(root1, root2)
}

fun root_str(r: Complex): String {
    if (r.im == 0.0) {
        return _numToStr(r.re)
    }
    var s: String = _numToStr(r.re)
    if (r.im >= 0.0) {
        s = ((s + "+") + _numToStr(r.im)) + "i"
    } else {
        s = (s + _numToStr(r.im)) + "i"
    }
    return s
}

fun user_main(): Unit {
    var roots: MutableList<Complex> = quadratic_roots(5.0, 6.0, 1.0)
    if (roots.size == 2) {
        println((("The solutions are: " + root_str(roots[0]!!)) + " and ") + root_str(roots[1]!!))
    }
}

fun main() {
    user_main()
}
